### PR TITLE
Pr/styled components layout box

### PR DESCRIPTION
**Pull request by niklaushug**

### CODE REVIEW

File: packages/tocco-ui/src/LayoutBox/StyledLayoutBox.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import styled from 'styled-components'
+import {theme} from 'styled-system'
+
+const declareBreakpoints = props => {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/LayoutBox/StyledLayoutBox.js]: const declareBreakpoints = (containerWidth, maxCellsPerRow) => {

  const xy = containerWidth <= 500 ?
    maxCellsPerRow.sm :
    containerWidth <= 1000 ?
      maxCellsPerRow.md :
      containerWidth <= 1500 ?
        maxCellsPerRow.md :
        maxCellsPerRow.lg

  return `${100 / xy}%`
}


- no repetitive code
- params are more intuitive

File: packages/tocco-ui/src/LayoutBox/example.js [hunk 2]
```diff
@@ -1,21 +1,81 @@
 /* eslint no-console: 0 */
 import React from 'react'
-import LayoutBox from './'
-// real-import:import {LayoutBox} from 'tocco-ui'
+import LayoutContainer, {LayoutBox} from './'
 
+import Panel, {PanelBody, PanelFooter, PanelHeader} from '../Panel'
+import {H4, H5, P, Span} from '../Typography'
+
+const ExampleCell = () => {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/LayoutBox/example.js]: const ExampleCell = () =>    <div style={{backgroundColor: '#ddd'}}>Cell</div>

File: packages/tocco-ui/src/LayoutBox/example.js [hunk 3]
```diff
@@ -1,21 +1,81 @@
...
+        <Span>Body</Span>
+      </PanelBody>
+      <PanelFooter showToggler={false}>
+        <H5>Footer</H5>
+      </PanelFooter>
+    </Panel>
+  )
+}
+
+// real-import:import {LayoutBox, LayoutContainer} from 'tocco-ui'
```
Comment (MethodenMann) [inline packages/tocco-ui/src/LayoutBox/example.js]: why is the "real-import" not on top?

File: packages/tocco-ui/src/Panel/Panel.js [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
 
+import StyledPanel from './StyledPanel'
+
 /**
- * <Panel/> is used to conceal and display related content alternating by interaction.
+ * <Panel/> is used to conceal and display related content alternating by interaction and to
+ * emphasize close relationship of content.
  */
 class Panel extends React.Component {
   state = {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Panel/Panel.js]: is this needed? I think a React.Component class has a empty state by default

File: packages/entity-detail/src/components/DetailForm/DetailForm.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ import PropTypes from 'prop-types'
 import React from 'react'
 import {reduxForm} from 'redux-form'
 import {intlShape, FormattedRelative, FormattedMessage} from 'react-intl'
-import {Button, LayoutBox} from 'tocco-ui'
+import Button from 'tocco-ui/src/Button'
```
[the remark this answers is not in the document]
Comment (niklaushug) [reply packages/entity-detail/src/components/DetailForm/DetailForm.js]: depends on https://github.com/tocco/tocco-client/pull/504

File: packages/tocco-ui/src/LayoutBox/LayoutContainer.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+import PropTypes from 'prop-types'
+import React from 'react'
+
+import StyledLayoutContainer from './StyledLayoutContainer'
+
+/*
+  TODO
+  o remove props.alignment
```
[the remark this answers is not in the document]
Comment (niklaushug) [reply packages/tocco-ui/src/LayoutBox/LayoutContainer.js]: done

File: packages/tocco-ui/src/LayoutBox/LayoutContainer.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+ */
+class LayoutContainer extends React.Component {
+  state = {
+    containerWidth: undefined
+  }
+
+  // get device independent pixel
+  getWidth = () => {
+    this.setState({
+      containerWidth: `${Math.floor(this.node.offsetWidth / window.devicePixelRatio)}`
```
[the remark this answers is not in the document]
Comment (niklaushug) [reply packages/tocco-ui/src/LayoutBox/LayoutContainer.js]: done

File: packages/tocco-ui/src/Panel/Panel.js (hunk 4 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/Panel.js]: state = {
    isOpen: this.props.isOpen
  }

is needed because initial state is defined by a props.

Alternatively I could rewrite row 37 to `isOpen: this.state.isOpen || this.props.isOpen,`

File: packages/entity-detail/src/components/DetailForm/DetailForm.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (rzueger) [reply packages/entity-detail/src/components/DetailForm/DetailForm.js]: #504 is merged now

File: packages/tocco-ui/src/Panel/Panel.js [hunk 8]
```diff
@@ -18,31 +21,45 @@ class Panel extends React.Component {
...
   isOpen: false,
   isToggleable: true
 }
 
 Panel.propTypes = {
   children: PropTypes.node,
   /**
-   * Boolean to control if <PanelBody/> is initially opened. Default value is 'false'.
+   * Boolean to control if <Panel.Header/>, <Panel.Body/> and <Panel.Footer/> is initially opened.
+   * Default value is 'false'.
```
Comment (rzueger) [inline packages/tocco-ui/src/Panel/Panel.js]: - Really to control if [...] is "initially openend"? Why is the prop called "isFramed"?
- The default value on line 49 is `true`, not `false`

File: packages/tocco-ui/src/Panel/Panel.js [hunk 9]
```diff
@@ -18,31 +21,45 @@ class Panel extends React.Component {
...
+   */
+  isFramed: PropTypes.bool,
+  /**
+   * Boolean to control if <Panel.Body/> is initially opened. Default value is 'true'.
```
Comment (rzueger) [inline packages/tocco-ui/src/Panel/Panel.js]: The default value on line 50 is `false`

File: packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
+
+import React from 'react'
+import {shallow} from 'enzyme'
+
+import PanelHeaderFooter from './PanelHeaderFooter'
+import Button from '../Button'
+import StyledPanelHeaderFooter from './StyledPanelHeaderFooter'
+
+describe('tocco-ui', function() {
```
Comment (rzueger) [inline packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js]: should there be an additional namespace `Panel` here so that it reflects the actual component structure?
--> some intermediate namespaces missing in some more tests too

File: packages/tocco-ui/src/Preview/Preview.js [hunk 11]
```diff
@@ -10,48 +10,60 @@ import {validateCssDimension} from '../utilStyles'
...
-  const onClick = () => {
-    if (typeof props.onClick === 'function') {
-      props.onClick(props.srcUrl, props.thumbnailUrl)
+class Preview extends React.Component {
+  state = {
+    loaded: false
+  }
+
+  handleOnClick = () => {
+    if (typeof this.props.onClick === 'function') {
```
Comment (rzueger) [inline packages/tocco-ui/src/Preview/Preview.js]: in my opinion we could shorten this to `if (this.props.onClick)`, since the prop type already checks if it's a function.

what do you think, @MethodenMann ?

File: packages/tocco-util/src/form/formBuilder.js [hunk 12]
```diff
@@ -136,18 +136,31 @@ export default (
...
+        output
+        = <Panel.Wrapper isFramed={true} isOpen={true}>
+            <Panel.Header><Typography.H4>{field.label}</Typography.H4></Panel.Header>
+            <Panel.Body>{output}</Panel.Body>
+          </Panel.Wrapper>
+      }
+
+      if (type === layoutTypes.HORIZONTAL_BOX) {
+        output = <Layout.Container key={key}>{output}</Layout.Container>
+      } else {
```
Comment (rzueger) [inline packages/tocco-util/src/form/formBuilder.js]: maybe handle all layout types explicitly and warn if the layout type is unknown?

File: packages/tocco-ui/src/Panel/StyledPanelBody.js [hunk 13]
```diff
@@ -1,8 +1,22 @@
 import styled from 'styled-components'
+import {theme} from 'styled-system'
 
 const StyledPanelBody = styled.div`
-  height: ${props => props.isOpen ? props.heightIfOpen : '0'};
-  overflow-y: auto;  /* if calculation does not work properly content is still accessible */
-  transition: height 300ms ease-in-out;
+  && {
+    height: ${props => props.isOpen ? props.heightIfOpen : '0'};
+    overflow-y: auto;  /* if calculation does not work properly content is still accessible */
```
Comment (rzueger) [inline packages/tocco-ui/src/Panel/StyledPanelBody.js]: the scrollbar is shown during the animation which doesn't look very nice in my opinion. should we make sure that the scrollbar isn't shown during the animation? what do you think?

File: packages/tocco-ui-showcase/src/components/Navigation.js [hunk 14]
```diff
@@ -30,7 +30,7 @@ const Navigation = props => {
           props.componentsTree.map(componentGroup => {
             return (
               <li key={componentGroup.category}>
-                <a href={'#' + componentGroup.category}>{componentGroup.category}</a>
+                <a href={'#category-' + componentGroup.category}>{componentGroup.category}</a>
```
Comment (MethodenMann) [inline packages/tocco-ui-showcase/src/components/Navigation.js]: use template literals

File: packages/tocco-ui-showcase/src/components/ShowCaseList.js [hunk 15]
```diff
@@ -10,7 +10,7 @@ const ShowCaseList = props => {
       {
         props.componentsTree.map(componentGroup => {
           return (
-            <section key={componentGroup.category} id={componentGroup.category}>
+            <section key={componentGroup.category} id={'category-' + componentGroup.category}>
```
Comment (MethodenMann) [inline packages/tocco-ui-showcase/src/components/ShowCaseList.js]: use template literals

File: packages/tocco-ui/src/Layout/Layout.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import React from 'react'
+
+/**
+ * Use <Layout.Container/> to apply grid and gutter on <Layout.Box>.
+ * Define how many cells are displayed per row and breakpoint.
+ */
+class Layout extends React.Component {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/Layout.js]: This is a dummy component for the showcase right? Please add as Comment to file

File: packages/tocco-ui/src/Layout/LayoutContainer.js [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.setDeviceIndependentPixelWidth)
+  }
+
+  render() {
+    return (
+      <StyledLayoutContainer
+        innerRef={node => { this.node = node }}
+        isNestedCorrectly={this.props.parent !== 'LayoutContainer'}
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/LayoutContainer.js]: Move keyword/magic string 'LayoutContainer' to a const?

File: packages/tocco-ui/src/Layout/StyledLayoutBox.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    : containerWidth <= 1000
+      ? maxCellsPerRow.md
+      : containerWidth <= 1500
+        ? maxCellsPerRow.lg
+        : maxCellsPerRow.xl
+  return `${100 / columns}%`
+}
+
+const declareStylesIfNestedCorrectly = props => {
+  if (props.isNestedCorrectly) {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/StyledLayoutBox.js]: use ternary operator?

File: packages/tocco-ui/src/Preview/Preview.js (hunk 11 above)
Comment (MethodenMann) [reply packages/tocco-ui/src/Preview/Preview.js]: 👍

File: packages/tocco-ui/src/Panel/Panel.js (hunk 8 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/Panel.js]: Descriptions were wrong.

File: packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js (hunk 10 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js]: Adding namespaces is not mandatory, because the component is imported directly and not resolved from index.js.

I admit it would enhance clarity if any example and unit test in tocco-ui would use dot notation (e.g. <Panel.Header> instead of <PanelHeaderFooter>). Such correction is a good task for the next quality day.

File: packages/tocco-util/src/form/formBuilder.js (hunk 12 above)
Comment (niklaushug) [reply packages/tocco-util/src/form/formBuilder.js]: see commit bd0fca475d99cce820e75330d2095f459cbf14a7

File: packages/tocco-ui/src/Panel/StyledPanelBody.js (hunk 13 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/StyledPanelBody.js]: I agree. see c5ae6d074a895b0110a716a30acb4b22ca7a1eb2

File: packages/tocco-ui/src/Panel/Panel.js (hunk 9 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/Panel.js]: see d78f73ba000fd74a33f13ce930f8291c78d87c09

File: packages/tocco-util/src/form/formBuilder.js [hunk 19]
```diff
@@ -135,20 +135,33 @@ export default (
...
+          <Panel.Body>{output}</Panel.Body>
+        </Panel.Wrapper>
+    }
+
+    if (type === layoutTypes.HORIZONTAL_BOX) {
+      output = <Layout.Container key={key}>{output}</Layout.Container>
+    } else if (type === layoutTypes.VERTICAL_BOX) {
+      output = <Layout.Box key={key}>{output}</Layout.Box>
+    } else {
+      console.warn(`Layout type "${type}" for box "${field.id}" is unknown.`) // eslint-disable-line no-console
```
Comment (rzueger) [inline packages/tocco-util/src/form/formBuilder.js]: use tocco-util/src/consoleLogger/consoleLogger.js

File: packages/tocco-util/src/form/formBuilder.js [hunk 20]
```diff
@@ -135,20 +135,33 @@ export default (
   }
 
   const createLayoutComponent = (field, type, key, traverser) => {
-    if (type === layoutTypes.HORIZONTAL_BOX || type === layoutTypes.VERTICAL_BOX) {
-      const alignment = type === layoutTypes.HORIZONTAL_BOX ? 'horizontal' : 'vertical'
-      const children = traverser()
+    let output = traverser()
```
[the remark this answers is not in the document]
Comment (niklaushug) [reply packages/tocco-util/src/form/formBuilder.js]: I've tried to do no reassignments and name variables better (see d8e66ed21e08f60eb244fefd9ef30bca16d52a12). I am not happy with the result because I could not achieve neither in a reasonable amout of time. May give me a hint.

File: packages/tocco-ui/src/Layout/LayoutContainer.js [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+import StyledLayoutContainer from './StyledLayoutContainer'
+import {layoutPropTypes, stylingLayout} from '../utilStyles'
+
+/**
+ * Use <Layout.Container/> to apply grid and gutter on <Layout.Box>.
+ * Define how many cells are displayed per row and breakpoint.
+ */
+class LayoutContainer extends React.Component {
+  state = {
+    containerWidth: undefined
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/LayoutContainer.js]: isnt undefined the default? I think you can remove this?

File: packages/tocco-ui/src/Layout/StyledLayoutContainer.js [hunk 22]
```diff
@@ -0,0 +1,29 @@
+import styled from 'styled-components'
+import {theme} from 'styled-system'
+
+const declareStylesIfNestedCorrectly = props => {
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/StyledLayoutContainer.js]: const declareStylesIfNestedCorrectly = props => 
  props.isNestedCorrectly ? 
    `
      display: flex;
      flex-wrap: wrap;
      margin: 0 -${theme('space.4')(props)};
      margin: 0 -15px;
       &:last-child {
        margin-bottom: 0;
      }
    `
  :
    `
      display: flex;
      flex-wrap: wrap;
      width: 100%;
    `
  }
}


would be shorter, but thats just a detail

File: packages/tocco-ui/src/Layout/example.js [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+import Panel from '../Panel'
+import Typography from '../Typography'
+import Layout from './'
+// real-import:import {Layout} from 'tocco-ui'
+
+const ExampleCell = () => <div style={{backgroundColor: '#ddd'}}>Cell</div>
+
+const ExamplePanel1 = () => {
+  return (
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Layout/example.js]: ) => <Panel.Wrapper>....

File: packages/tocco-ui/src/Panel/Panel.js [hunk 24]
```diff
@@ -18,31 +21,45 @@ class Panel extends React.Component {
...
+
     return (
-      <div>
+      <StyledPanel
+        isFramed={isFramed}
+      >
         {
           React.Children.map(this.props.children, child =>
             React.cloneElement(child, {
+              isFramed: isFramed,
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Panel/Panel.js]: isFramed,

works with ES6

File: packages/tocco-ui/src/Panel/Panel.js [hunk 25]
```diff
@@ -18,31 +21,45 @@ class Panel extends React.Component {
...
               isOpen: this.state.isOpen,
-              isToggleable: this.props.isToggleable,
+              isToggleable: isToggleable,
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Panel/Panel.js]: same here

File: packages/tocco-ui/src/Panel/PanelBody.js [hunk 26]
```diff
@@ -4,43 +4,66 @@ import React from 'react'
 import StyledPanelBody from './StyledPanelBody'
 
 /**
- * Only <PanelBody/> is affected by the visibility state.
+ * Only <Panel.Body/> is affected by the visibility state.
  */
 class PanelBody extends React.Component {
+  observer = null
```
Comment (MethodenMann) [inline packages/tocco-ui/src/Panel/PanelBody.js]: i dont think this initialization with null is necessary?

same applies vor the state?

File: packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js (hunk 10 above)
Comment (MethodenMann) [reply packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js]: hm rzueger is right. 

`describe('Panel', ....`

is missing. We we have the whole folder structure in our spec files. It doenst matter how a component is imported.

File: packages/tocco-util/src/form/formBuilder.js [hunk 27]
```diff
@@ -135,19 +136,30 @@ export default (
...
+    if (Array.isArray(elements)) {
+      elements = elements.filter(el => el)
+    }
 
-      return (
-        <LayoutBox key={key} label={field.label} alignment={alignment}>
-          {children}
-        </LayoutBox>
-      )
+    if (!elements || (Array.isArray(elements) && elements.every(e => e == null))) {
```
Comment (MethodenMann) [inline packages/tocco-util/src/form/formBuilder.js]: ` && elements.every(e => e == null)`

makes no sense here because on line 142 you filtered out any element that is null.

`&& elements.length === 0`

File: packages/tocco-util/src/form/formBuilder.js [hunk 28]
```diff
@@ -135,19 +136,30 @@ export default (
...
+    }
+
+    const content = field.label
+      ? <Panel.Wrapper isFramed={true} isOpen={true}>
+        <Panel.Header><Typography.H4>{field.label}</Typography.H4></Panel.Header>
+        <Panel.Body>{elements}</Panel.Body>
+      </Panel.Wrapper>
+      : elements
+
+    if (type === layoutTypes.HORIZONTAL_BOX) {
```
Comment (MethodenMann) [inline packages/tocco-util/src/form/formBuilder.js]: why not:

```
const layoutMap = {
   [layoutTypes.HORIZONTAL_BOX]: Layout.Container,
   [layoutTypes.VERTICAL_BOX]: Layout.Box
}
const LayoutComponent = layoutMap[type] 

if (LayoutComponent) {
  return <LayoutComponent key={key}>{content} </LayoutComponent >
} else {
      consoleLogger.logWarning(`Layout type "${type}" for box "${field.id}" is unknown.`)
      return <div key={key}>{content}</div>
 }	    

```
But this only if its a realistic scenario that "type" is neither a horizontal box nor a vertical box.
Otherwise:

```

const LayoutComponent = type === layoutTypes.HORIZONTAL_BOX ? Layout.Container : Layout.Box
 return <LayoutComponent key={key}>{content} </LayoutComponent >

```

File: packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js (hunk 10 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Panel/PanelHeaderFooter.spec.js]: I missed the point. Thanks for being persistent.

File: packages/tocco-ui/src/Layout/LayoutContainer.js (hunk 21 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Layout/LayoutContainer.js]: I still need to declare `state = {}` in order that `containerWidth: this.state.containerWidth` does not fail on line 38.

File: packages/tocco-ui/src/Layout/StyledLayoutContainer.js (hunk 22 above)
Comment (niklaushug) [reply packages/tocco-ui/src/Layout/StyledLayoutContainer.js]: Personally, I prefer readable over compact css. Build process shall remove unnecessary whitespace, which it does most likely already.

Comment (MethodenMann) [reply packages/tocco-ui/src/Layout/StyledLayoutContainer.js]: dont get me wrong. its not about the css (all on one line happend when i copy&paste and should not be done).
Its about replaceing

const xy = props => {
if (x) {return y} else {return z}
}

with const xy = props =>
  x? y : z
